### ISSUE TITLE
StackOverflow when creating or utilising spec-tools with self referencing spec

### ISSUE THREAD

**Issue by polvoblanco** [metosin/spec-tools#244]
e.g.
`(s/def ::pattern (s/coll-of (s/or :attr qualified-keyword?
                                  :ref (s/map-of qualified-keyword?
                                                 (s/or :pattern ::pattern
                                                       :recur #{:...})))
                            :kind vector? :min-count 1 :distinct true))`

Works as expected in vanilla spec, in spec-tools however, calling st/spec or st/select-spec etc. causes a StackOverflow error in parse.cljc.

**Comment by wandersoncferreira** [comment]
Hello @polvoblanco , how are you? I looked into your reported problem and indeed the `parse-spec` function is not handling recursive calls. I added an PR at #246 with a possible solution to this problem and added a test simulating the desired scenario.

If you would be willing to pull from my branch to test the implementation further would be great. Thanks for reporting and providing an example to reproduce the error.
